### PR TITLE
types/unist: Change unist types

### PR DESCRIPTION
**Pull request by wooorm**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/syntax-tree/unist/pull/98
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

Update for [unist 3](https://github.com/syntax-tree/unist/releases/tag/3.0.0) (removal of `indent` field) and remove generics

##### Change

*   Remove generic type parameter of `unist` `Node`, `Literal`, `Parent`
*   Remove `indent` of `unist` `Position`
    (see https://github.com/syntax-tree/unist/releases/tag/3.0.0)
*   Change to make `hast` `Element` `properties` required
*   Remove `name` field on `hast` `DocType`, which must always be `'html'`
    (see https://github.com/syntax-tree/hast/releases/tag/2.4.0)
*   Rename `hast` `DocType` to `Doctype`.
*   Remove `mdast` `Footnote` node, which is not used by GFM
    (see https://github.com/syntax-tree/mdast/releases/tag/5.0.0)
*   Rename `mdast` `HTML` to `Html` and `YAML` to `Yaml`
    (see https://github.com/syntax-tree/mdast/releases/tag/5.0.0)
*   Change to make `xast` `Element` `attributes` required

##### Add

*   Add data registries to all nodes
*   Add types for different concrete node unions (such as `Literals` from
    for all literal xast nodes, `Nodes` for all nodes, etc.)

##### Refactor

*   Add lots of doc comments explaining everything

#### Note: generics

Generics were added in GH-54413. This commit undoes generics.

I was neutral on generics 2 years ago, but now that our entire ecosystem is typed, I am against.

There were three generics:

*   `children` and `value`:
    unist is supposed to be an abstract interface that has to be extended before being used.
    *Those* types are supposed to choose a `type` field, specify which other fields exist, and specify particular children or value.
*   `data`:
    I don’t think generic data is a good idea. `Data` is a *shared* space that *several* plugins/utilities can access to store whatever they want. A generic makes it possible to hide conflicts.
    Generics also hide what is actually supported.

    We’ve been adding an alternative in other places (micromark, vfile, mdast/hast/etc for content types of nodes), and I think we should do the same for data: use an interface that can be registered onto.

    While data is not useful on `unist`, I plan to do the same to `mdast` and `hast`, where it gets very useful: `mdast-util-to-hast` for example sets a [`meta` field on `hast`s `element.data`](https://github.com/syntax-tree/mdast-util-to-hast/blob/c87cd606731c88a27dbce4bfeaab913a9589bf83/lib/handlers/code.js#L41), and it adds support for `hProperties`/`hChildren`/`hName` on all mdast nodes. If that utility registers with `hast` `Element.Data` and `mdast` `Node.data`, all other utilities and plugins can get typed data.

Lastly, I searched GH for `unist` and `Node<` and did not see much use, expect for  ugly generated types, such as: https://github.com/nachoaldamav/hardlinks-test/blob/f40b5caf13a53dd063fa5cd0aa7731e73b12631d/data/unist-util-map/index.d.ts#L33.

### CODE REVIEW

**Review by remcohaszing — APPROVED** [review]
Feel free to add me to the list of owners, @typescript-bot will ping me 🙂

**Comment by wooorm** [comment]
Thanks for the review. Note also my Q:

> Q: how can I bundle several major changes in several PRs together?

**Comment by wooorm** [comment]
a) it breaks the tests
b) I want to do more breaking things!

**Comment by remcohaszing** [comment]
Technically this could be considered another breaking change for `vfile` / `vfile-message` as well then.

File: types/unist/index.d.ts
Comment (wooorm) [inline types/unist/index.d.ts:52]: I specifically removed the record fallback. As we’re in TS, the goal is for everyone to type everything.

My goal is for an `hast.Data` to exist which includes all of `unist.Data`, same for hast/xast/nlcst

**Comment by wooorm** [comment]
The CI is supposed to break. This is a major breaking change.

File: types/unist/index.d.ts
Comment (remcohaszing) [inline types/unist/index.d.ts:12]: I believe the idiomatic thing is to omit this line.

```suggestion
```

File: types/unist/index.d.ts
Comment (remcohaszing) [reply types/unist/index.d.ts:52]: I have my doubts about this one. Is there ever going to be a data property that is applicable to all unist nodes?

Comment (wooorm) [reply types/unist/index.d.ts:51]: I could imagine an `id`, sort of like a `key` from VDOMs.

One thing I’ve been vaguely wondering about is how nodes could come from different files, nodes could have some info leading back to which file they came from?

Regardless, we already have data, so what do you propose? Removing it?
We have interfaces everywhere. The point is to remove the generics and implement this in mdast/hast/xast/nlcst and utilities!

**Comment by wooorm** [comment]
The CI is still supposed to break :) This is a major breaking change.

File: types/unist/index.d.ts
Comment (remcohaszing) [reply types/unist/index.d.ts:51]: I was fine with the solution we had using generics, but also with just specifying it as `Record<string, unknown>` and removing the generic and this type.

Nodes that extend `Node` can just override the `data` property. E.g.:

```ts
interface MdxEsm extends Node {
  type: 'mdxEsm',

  data: {
    program: Program
  }
}
```

I’m leaning towards just removing it.

Comment (wooorm) [reply types/unist/index.d.ts:51]: Please see [my commit message for more info](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/65789/commits/e8f0962465519d2ef410da2d6ffa2156135b2ef7). Your proposal does not work with utilities that add to nodes, such as the use cases for `mdast-util-to-hast`.

Comment (wooorm) [reply types/unist/index.d.ts:51]: @remcohaszing Ping! :)

Comment (remcohaszing) [reply types/unist/index.d.ts:51]: If we have a use case for adding data to unist nodes globally, I’m all for having the interface the way you have defined it right now. :)

Note that it’s also possible to combine generics with this solution.

```ts
// unist

interface Node<T = object> {
  type: string
  data: Data & T
}

interface Data {}

// mdast-util-to-hast

interface Data {
  hName: string
}

// mdx

interface MdxEsm extends Node<{ estree: {type: 'Program'} }> {
  type: 'mdxEsm'
}
```

as opposed to:

```ts
// unist

interface Node {
  type: string
  data: Data
}

interface Data {}

// mdast-util-to-hast

interface Data {
  hName: string
}

// mdx

interface MdxEsmData extends Data {
  estree: {type: 'Program'}
}

interface MdxEsm extends Node {
  type: 'mdxEsm'
  data: MdxEsmData
}
```

Although the latter isn’t too bad either.

Comment (wooorm) [reply types/unist/index.d.ts:51]: I indeed like the latter. It supports the needs of `mdast-util-to-hast`, which as far as I understand generics do not.

I want to make a chain. EsmData <- MdastData <- UnistData.

`UnistData` is probably not particularly useful, but it is specified: https://github.com/syntax-tree/unist#data.
Two example use cases are in  https://github.com/DefinitelyTyped/DefinitelyTyped/pull/65789#discussion_r1231098799.

But I think it’s the same as `Node` from unist: that is never used either.
It’s only used in the implementation chain: EsmNode <- MdastNode <- UnistNode.
We don’t type `Node` as `Record<string, unknown>` either?

**Review by remcohaszing — APPROVED** [review]

**Comment by remcohaszing** [comment]
This has my approval, but that's not sufficient until this is merged. :)

Also dependant types need to be adjusted to fix CI.

**Review by ChristianMurphy — APPROVED** [review]

**Comment by wooorm** [comment]
> but that's not sufficient until this is merged. :)
>
> Also dependant types need to be adjusted to fix CI.

Do you mean I should patch the breaking changes in one PR to get them released?

**Comment by remcohaszing** [comment]
> Do you mean I should patch the breaking changes in one PR to get them released?

Yes. I suppose this also means mdast, hast, and xast should get semver major releases?

Alternatively you can track types for multiple versions of unist. I.e. you can have a look at `react` to see how that’s done.

**Review by ChristianMurphy — APPROVED** [review]

File: types/mdast/index.d.ts
Comment (remcohaszing) [inline types/mdast/index.d.ts:684]: This line occurs multiple times. Maybe it’s better to just disable this tslint rule entirely.

Comment (wooorm) [reply types/mdast/index.d.ts]: done!

**Review by remcohaszing — APPROVED** [review]

**Review by ChristianMurphy — APPROVED** [review]

**Review by navya9singh — APPROVED** [review]

**Comment by wooorm** [comment]
Ready to merge

**Comment by Methuselah96** [comment]
@wooorm Not sure where best to report this, but I'm getting type errors due to the fact that `@types/mdast@3.0.11` has a dependency on `@types/unist@*` (since DefinitelyTyped uses `*` by default) which ends up resolving to `@types/unist@3.0.0` instead of `@types/unist@2`.

My lock file looks something like this:

```
"@types/mdast@npm:^3.0.0":
  version: 3.0.11
  resolution: "@types/mdast@npm:3.0.11"
  dependencies:
    "@types/unist": "*"
  checksum: 3b04cf465535553b47a1811c247668bd6cfeb54d99a2c9dbb82ccd0f5145d271d10c3169f929701d8cd55fd569f0d2e459a50845813ba3261f1fb0395a288cea
  languageName: node
  linkType: hard

"@types/unist@npm:*":
  version: 3.0.0
  resolution: "@types/unist@npm:3.0.0"
  checksum: e9d21a8fb5e332be0acef29192d82632875b2ef3e700f1bc64fdfc1520189542de85c3d4f3bcfbc2f4afdb210f4c23f68061f3fbf10744e920d4f18430d19f49
  languageName: node
  linkType: hard

"@types/unist@npm:^2.0.0":
  version: 2.0.6
  resolution: "@types/unist@npm:2.0.6"
  checksum: 25cb860ff10dde48b54622d58b23e66214211a61c84c0f15f88d38b61aa1b53d4d46e42b557924a93178c501c166aa37e28d7f6d994aba13d24685326272d5db
  languageName: node
  linkType: hard
```

Let me know if you need any more details or can help out at all.

**Comment by ChristianMurphy** [comment]
@Methuselah96 see https://github.com/remarkjs/remark-inline-links/issues/7#issuecomment-1627633383
Quoting @remcohaszing from that comment:

---

This issue is caused by `@types/mdast` depending on `@types/unist@*`. Unfortunately DefinitelyTyped uses `*` dependency ranges. This issue exists now that we're in the process of updating all packages in the unified ecosystem.

You can solve the issue using npm resolutions or yarn overrides in `package.json`.

```jsonc
{
  // for npm
  "resolutions": {
    "@types/node": "2"
  },
  // for yarn
  "overrides": {
    "@types/node": "2"
  }
}
```

For pnpm there may be a similar solution, but I'm not sure what. This is a temporary workaround. The issue should resolve itself later.

**Comment by wooorm** [comment]
@ChristianMurphy there’s a typo in the code: `types/node` is not needed, `types/unist` is

---

> Let me know if you need any more details or if I can help out at all.

Thanks! I don’t think there’s much one can do other than the resolution.
And we should removing the `*`s from the types here (and perhaps we need to search for other types here that depend on `@types/{unist,mdast,hast,...}` as `*`?)

**Comment by Methuselah96** [comment]
I think we could fix the `@types/mdast@3` dependency on `@types/unist@*` by creating a folder for the earlier major versions and have one depend on the other (as discussed [here](https://github.com/DefinitelyTyped/DefinitelyTyped#if-a-library-is-updated-to-a-new-major-version-with-breaking-changes-how-should-i-update-its-type-declaration-package)) and then delete them right after if we don't care about maintaining the older major versions. I could create a PR demonstrating what it would look like if you're interested. This is the process that I follow for the d3 types and it seems to work sufficiently.

**Comment by wooorm** [comment]
If that works, nice, yeah! Would need the same for hast/nlcst/xast, too!

**Comment by Methuselah96** [comment]
@wooorm https://github.com/DefinitelyTyped/DefinitelyTyped/pull/66029 should do the trick.

**Pull request merged**
